Bluetooth: Controller: Implement bt_hci_driver::close

carlescufi opened the pull request:
The open source controller did not expose support for shutting it down
via the close() API in the HCI driver. Add support for it, which
completes the support for disabling and enabling the Bluetooth stack in
Zephyr.

Closes #3192.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
Bluetooth: Controller: Implement bt_hci_driver::close

The open source controller did not expose support for shutting it down
via the close() API in the HCI driver. Add support for it, which
completes the support for disabling and enabling the Bluetooth stack in
Zephyr.

Closes #3192.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/hci/hci_driver.c:
```diff
@@ -63,8 +63,8 @@
 
-static K_SEM_DEFINE(sem_prio_recv, 0, K_SEM_MAX_LIMIT);
-static K_FIFO_DEFINE(recv_fifo);
+static struct k_sem sem_prio_recv;
+static struct k_fifo recv_fifo;
```
Why were the static initializers removed in favor of `k_fifo_init` and `k_sem_init`?

carlescufi replied:
Initializing a FIFO makes it empty. Initializing a semaphore resets its count to 0. Since the HCI driver `close()` function can be called at any time, we do need to reset these structures every time `open()` is called (or once at boot and then every time `close()` is called.

>Does it work as intended if hci_driver_open is called multiple times? 

You are not supposed to call `open()` multiple times without calling `close()` first. But if you do that, this will make it work as intended, since we reset all the structs at that point.

ppryga-nordic reviewed (dismissed):
LGTM

cvinayak commented on subsys/bluetooth/controller/hci/hci_driver.c:
```diff
@@ -734,13 +733,17 @@ static int hci_driver_open(void)
+	k_fifo_init(&recv_fifo);
+	k_sem_init(&sem_prio_recv, 0, K_SEM_MAX_LIMIT);
+
 	err = ll_init(&sem_prio_recv);
```
`lll_init()` requests LF clock on, multiple `open()` could accumulate the reference count in the clock_control driver?

Either `lll_init()` has to be moved outside of `open()` to be only at power on reset or introduce a `lll_deinit()` which is a combination of `lll_reset()` plus a new `lll_clock_deinit()`  so that LF clock request is released.

Thalley replied:
If something needs to be run just once, could they be put in `hci_driver_init`?

carlescufi replied:
>Either lll_init() has to be moved outside of open() to be only at power on reset or introduce a lll_deinit() which is a combination of lll_reset() plus a new lll_clock_deinit() so that LF clock request is released.

If we did this, then `open()` would do nothing LL-related right?

cvinayak replied:
> If we did this, then `open()` would do nothing LL-related right?

Yes, and calls to only `ll_reset()` will suffice inside the `close` interface.

carlescufi committed:
Bluetooth: Controller: Add deinit() infrastructure

Allow the controller to be deinitialized, adding the whole chain of
calls:

- ll_deinit()
- lll_deinit()
- lll_clock_deinit()

in order to be able to turn everything off, including the controller's
refcount of the LF clock.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll.c:
```diff
@@ -212,6 +212,28 @@ int lll_init(void)
+	irq_disable(RADIO_IRQn);
+	irq_disable(RTC0_IRQn);
+	irq_disable(HAL_SWI_RADIO_IRQ);
+#if defined(CONFIG_BT_CTLR_LOW_LAT) || \
+	(CONFIG_BT_CTLR_ULL_HIGH_PRIO != CONFIG_BT_CTLR_ULL_LOW_PRIO)
+	irq_disable(HAL_SWI_JOB_IRQ);
+#endif
```
Isn't `HAL_SWI_JOB_IRQ` always defined? If so, then we can replace the `#if defined` with a `IS_ENABLED`

carlescufi replied:
Yep, fine with that. @cvinayak you OK with this?

cvinayak replied:
yes.

cvinayak commented:
I get segmentation fault when using `bt_disable()` in a modified `samples/bluetooth/broadcaster` sample.

![image](https://user-images.githubusercontent.com/6350656/171663585-77ca66b9-64c5-4689-9315-fc0d2d74892c.png)

![image](https://user-images.githubusercontent.com/6350656/171663943-95a53c45-547a-46d9-81de-f571e65f5972.png)

carlescufi committed:
Bluetooth: Host: Properly handle disable()/enable() cycles

When bt_disable() was introduced, some of the global variables that
handle key aspects of Bluetooth were not converted to be also
re-initialized every time that the stack was disabled and the enabled
again. This meant that the FIFO and semaphore in the device structure
were not reset whenever the corresponding threads using them were shut
down and then restarted.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

cvinayak commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -2455,7 +2445,7 @@ static void process_events(struct k_poll_event *ev, int count)
 static void hci_tx_thread(void *p1, void *p2, void *p3)
 {
-	static struct k_poll_event events[EV_COUNT] = {
+	struct k_poll_event events[EV_COUNT] = {
```
Not sure by how much the `hci_tx_thread`'s program stack size needs increase, if retaining the `static` does not fault, lets keep it back.

carlescufi replied:
Seems to work fine with the `static`, so I'm bringing it back.

cvinayak commented on subsys/bluetooth/controller/ll_sw/openisa/lll/lll.c:
```diff
@@ -166,6 +166,20 @@ int lll_init(void)
+	irq_disable(LL_RADIO_IRQn);
+	irq_disable(LL_RTC0_IRQn);
+	irq_disable(HAL_SWI_RADIO_IRQ);
+#if defined(CONFIG_BT_CTLR_LOW_LAT) || \
```
Use IS_ENABLE here too? (I have not checked if these are the correct IRQs for openisa port...)

carlescufi replied:
Fixed

cvinayak commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3655,6 +3645,16 @@ int bt_enable(bt_ready_cb_t cb)
+	 * exception is if the controller requests to wait for an
+	 * initial Command Complete for NOP.
+	 */
+#if !defined(CONFIG_BT_WAIT_NOP)
```
Use IS_ENABLED here too...

carlescufi replied:
Fixed

carlescufi commented:
@ppryga @Thalley @cvinayak please take another look

asbjornsabo reviewed (approved):
LGTM

ppryga-nordic reviewed (approved):
LGTM
